### PR TITLE
bpo-37702: Fix SSL certificate-store-handles leak

### PR DESCRIPTION
**Pull request by neonene**
In Windows, CertCloseStore(hCollectionStore,[any option]) closes only hCollectionStore.
To avoid out-of-memory, CertCloseStore() shuld be called to each hSystemStore
which was added to hCollectionStore.

I think CERT_CLOSE_STORE_FORCE_FLAG is rare option and not much needed.
MSDN says CertCloseStore(handle, 0) is ideal and works fine for me.

https://bugs.python.org/issue37702

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](https://bugs.python.org/) (b.p.o) to verify you have signed the CLA (this might be simply due to a missing "GitHub Name" entry in your b.p.o account settings). This is necessary for legal reasons before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for your contribution, we look forward to reviewing it!

**Comment by zooba** [comment]
The theory seems fine. I haven't checked the implementation - can't say when I'll get to it, but I'll try and do it before the next release.

File: Modules/_ssl.c
Comment (epicfaace) [inline Modules/_ssl.c]: You could remove the `BOOL success;` declaration before and do this?:
```suggestion
    BOOL success = CertCloseStore(hCollectionStore, CERT_CLOSE_STORE_FORCE_FLAG);
```

File: Misc/NEWS.d/next/Windows/2019-07-29-16-49-31.bpo-37702.Lj2f5e.rst
Comment (epicfaace) [inline Misc/NEWS.d/next/Windows/2019-07-29-16-49-31.bpo-37702.Lj2f5e.rst]: It might be good to specify which methods this applies to, and that it is only on Windows.

Comment (neonene) [reply Misc/NEWS.d/next/Windows/2019-07-29-16-49-31.bpo-37702.Lj2f5e.rst]: I'd rather not describe c-implementation level information.

File: Modules/_ssl.c
Comment (epicfaace) [inline Modules/_ssl.c]: Do you think it would be good to keep in the definition of CERT_CLOSE_STORE_FORCE_FLAG ("forces freeing of memory for all contexts associated with the store")?

Comment (neonene) [reply Modules/_ssl.c]: Thanks for Bool position.
For now, both 0 and CERT_CLOSE_STORE_FORCE_FLAG  work fine for my pc.
I'd like to keep changes to a minimum and leave it to other ssl experts' judgement whether the flag should revert to 0 or not.

Comment (epicfaace) [reply Modules/_ssl.c]: Oh, I didn't mean reverting the flag to 0, I meant, did you want to keep the part in the comment that says "CERT_CLOSE_STORE_FORCE_FLAG forces freeing of memory for all contexts associated with the store"?

Comment (neonene) [reply Modules/_ssl.c]: MS says,
CERT_CLOSE_STORE_FORCE_FLAG:
Forces the freeing of memory for all contexts associated with the store.
This flag can be safely used only when the store is opened in a function
and neither the store handle nor any of its contexts are passed to
any called functions. For details, see ...  

I think, keeping short sentences caused and will cause a misunderstanding that
this rarely used flag is always safe and releases all certificate-store-handles.
Having said that, I'm not in favor of long description in the source file.

**Comment by neonene** [comment]
MS  doesn't recommend using CERT_CLOSE_STORE_FORCE_FLAG.
https://blogs.msdn.microsoft.com/winsdk/2010/01/29/passing-the-flag-cert_close_store_force_flag-to-certclosestore-may-cause-your-application-to-crash/

And I don't want to leave confusing comments on the flag that is no longer necessary to close all handles.
So, I decided to change the flag back to 0.

**Comment by neonene** [comment]
I'm done. Sorry for a dirty PR.

File: Modules/_ssl.c
Comment (tiran) [inline Modules/_ssl.c]: I don't like code duplication. Please make the system store list a global static

File: Modules/_ssl.c
Comment (tiran) [inline Modules/_ssl.c]: This style is confusing and hard to read. Split ``size_t i`` and ``size_t system_stores_count`` into two lines.

File: Modules/_ssl.c
Comment (tiran) [inline Modules/_ssl.c]: The comment is out of date. You are referencing ``CERT_CLOSE_STORE_FORCE_FLAG`` but no longer using ``CERT_CLOSE_STORE_FORCE_FLAG`` as flag.

**Review by tiran — CHANGES_REQUESTED** [review]

**Comment by neonene** [comment]
Please check.
I'm relieved my flag change is not rejected.

**Comment by neonene** [comment]
I have made the requested changes; please review again.

**Comment by neonene** [comment]
Oh, I should have enclosed #define in brackets.

**Comment by neonene** [comment]
By any chance, Is this PR for Windows needs further review from zooba?
Thanks to tiran and epicfaace, I think it have become readable.

**Comment by neonene** [comment]
I raised another PR, which fix store leaks rather than handle leaks.
https://github.com/python/cpython/pull/15632

**Comment by zooba** [comment]
I'm closing this as it's unnecessary with the other fix (I verified there is no handle leak or memory leak with `urlopen` anymore). Thanks for your efforts! They're appreciated.

**Pull request closed without merging**
